pika/adapters/asyncio_connection.py

```diff
@@ -0,0 +1,476 @@
+    queue, ctag = yield from chan.basic_consume(queue=cb_queue, no_ack=True)
+
+    yield from chan.basic_publish(
+        exchange='',
+        routing_key='rpc_queue',
+        properties=pika.BasicProperties(
+             reply_to=cb_queue,
+             ),
+        body='Hello World!')
+
+    ch, method, props, body = yield from queue.get()
+    print(body)
+
+loop = asyncio.get_event_loop()
+
+try:
+    task = asyncio.ensure_future(client(loop))
+    loop.run_until_complete(task)
+except KeyboardInterrupt:
+    print('Done')
+
+
+################
+# RPC server
+################
+
+import asyncio
+import pika
+
+from pika.adapters import asyncio_connection
+
+
+@asyncio.coroutine
+def make_connection(loop, host="localhost", port=5672):
+
+    def connection_factory():
+        params = pika.ConnectionParameters()
+        return asyncio_connection.AsyncioProtocolConnection(params, loop=loop)
+
+    transport, connection = yield from loop.create_connection(connection_factory, host, port)
+    yield from connection.ready # important!
+    return connection
+
+
+@asyncio.coroutine
+def server(loop):
+    conn = yield from make_connection(loop)
+    chan = yield from conn.channel()
+
+    yield from chan.queue_declare(queue='rpc_queue')
+    queue, ctag = yield from chan.basic_consume(queue='rpc_queue', no_ack=True)
+
+    while True:
+        ch, method, props, body = yield from queue.get()
+        yield from chan.basic_publish(
+            exchange='',
+            routing_key=props.reply_to,
+            body=body[::-1])
+
+
+loop = asyncio.get_event_loop()
+
+try:
+    task = asyncio.ensure_future(server(loop))
+    loop.run_until_complete(task)
+except KeyboardInterrupt:
+    print('Done')
+
+"""
+import functools
+import asyncio
+import socket
+from logging import getLogger
+
+from pika import exceptions
+from pika.adapters import base_connection
+
+
+
+def _fail(failure=None):
+    f = asyncio.Future()
+    f.set_exception(failure)
+    return f
+
+
+def _succeed(res):
+    f = asyncio.Future()
+    f.set_result(res)
+    return f
+
+
+class ClosableQueue(asyncio.Queue):
+    """
+    Like the normal asycio Queue, but after close() is called with an
+    Exception instance all pending Futures are errbacked and further attempts
+    to call get() or put() return a Failure wrapping that exception.
+    """
+
+    def __init__(self, maxsize=0, *, loop=None):
+        self.closed = None
+        super().__init__(maxsize, loop=loop)
+
+    def put(self, obj):
+        if self.closed:
+            return _fail(self.closed)
+        return super().put(obj)
+
+    def get(self):
+        if self.closed:
+            return _fail(self.closed)
+        return super().get()
+
+    def close(self, reason):
+        self.closed = reason
+        while self._getters:
+            waiter = self._getters.popleft()
+            if not waiter.done():
+                waiter.set_exception(reason)
+        self._queue.clear()
+
+
+class AsyncioChannel:
```

Comment:
This should probably use modern-style class declaration and be derived from `object`.